Mathlib/Tactic/Linter/DirectoryDependency.lean

```diff
@@ -472,25 +547,75 @@ end DirectoryDependency
       if !overrideAllowedImportDirs.contains mainModule imported then
-        let mut msg := m!"Modules starting with {n₁} are not allowed to import modules starting with {n₂}.
-This module depends on {imported}\n"
+        let mut msg := m!"Modules starting with {n₁} are not allowed to import modules starting with {n₂}. \
+        This module depends on {imported}\n"
         for dep in env.importPath imported do
           msg := msg ++ m!"which is imported by {dep},\n"
-        return some <| msg ++ m!"which is imported by this module.
-(Exceptions can be added to `overrideAllowedImportDirs`.)"
+        return some (msg ++ m!"which is imported by this module. \
+          (Exceptions can be added to `overrideAllowedImportDirs`.)")
+      else none
     else
-      return some m!"Internal error in `directoryDependency` linter: this module claims to depend on a module starting with {n₂} but a module with that prefix was not found in the import graph."
-  | none => pure ()
-  return none
+      return some m!"Internal error in `directoryDependency` linter: this module claims to depend \
+      on a module starting with {n₂} but a module with that prefix was not found in the import graph."
+  | none => none
+
+@[inherit_doc Mathlib.Linter.linter.directoryDependency]
+def directoryDependencyCheck (mainModule : Name) : CommandElabM (Array MessageData) := do
+  unless Linter.getLinterValue linter.directoryDependency (← getOptions) do
+    return #[]
+  let env ← getEnv
+  let imports := env.allImportedModuleNames
+
+  -- If this module is in the allow-list, we only allow imports from directories specified there.
+  -- Collect all prefixes which have a matching entry.
+  let matchingPrefixes := mainModule.prefixes.filter (fun prf ↦ allowedImportDirs.containsKey prf)
+  if matchingPrefixes.isEmpty then
+    -- Otherwise, we fall back to the blocklist `forbiddenImportDirs`.
+    if let some msg := _checkBlocklist env mainModule imports then return #[msg] else return #[]
+  else
+    -- We always allow imports in the same directory (for each matching prefix),
+    -- and from `Init`, `Lean`, `Std` and `Qq`.
+    -- We also allow transitive imports of Mathlib.Init, as well as Mathlib.Init itself.
+    let initImports := (← findImports ("Mathlib" / "Init.lean")).append
+      #[`Mathlib.Init, `Mathlib.Tactic.DeclarationNames]
+    let exclude := [
+      `Init, `Std, `Lean, `Qq,
+    ]
+    let importsToCheck := imports.filter (fun imp ↦ !exclude.any (·.isPrefixOf imp))
+      |>.filter (fun imp ↦ !matchingPrefixes.any (·.isPrefixOf imp))
+      |>.filter (!initImports.contains ·)
+
+    -- Find all prefixes which are allowed for one of these directories.
+    let mut allRules := NameSet.empty
+    for prfix in matchingPrefixes do
+      let some rules := RBMap.find? allowedImportDirs prfix | unreachable!
+      allRules := allRules.append rules
+    -- Error about those imports which are not covered by allRules.
```

Comment:
This comment was a bit misleading to me at first, since the check is really `allowedImportDirs.contains`, and does not refer to `allRules`.

Reply:
Good catch, fixed.